server/src/client/server.rs

```diff
@@ -1,100 +1,110 @@
-    pub session: Addr<ClientWsSession>,
-    pub session_id: SessionId,
-    pub login: Login,
+#[derive(Debug, Message)]
+pub struct LogoutUserSessions {
+    pub user_id: UserId,
 }
 
-#[derive(Message)]
+#[derive(Debug, Message)]
 pub struct Disconnect {
-    pub session_id: SessionId,
-    pub user_id: Option<UserId>,
+    pub user_id: UserId,
+    pub device_id: DeviceId,
 }
 
-#[derive(Debug, Message)]
+#[derive(Debug)]
 pub struct Join {
     pub room: RoomId,
 }
 
+impl Message for Join {
+    type Result = RequestResponse;
+}
+
 impl ClientMessageType for Join {}
 
-impl MessageResponse<ClientServer, IdentifiedMessage<CreateRoom>> for RoomId {
-    fn handle<R: ResponseChannel<IdentifiedMessage<CreateRoom>>>(
-        self,
-        _: &mut Context<ClientServer>,
-        tx: Option<R>,
-    ) {
-        if let Some(tx) = tx {
-            tx.send(self)
-        }
-    }
+#[derive(Debug)]
+pub struct CreateRoom;
+
+impl ClientMessageType for CreateRoom {}
+
+impl Message for CreateRoom {
+    type Result = RequestResponse;
 }
 
 #[derive(Debug)]
 pub struct IdentifiedMessage<T: Message + ClientMessageType + Debug> {
-    pub session_id: SessionId,
     pub user_id: UserId,
+    pub device_id: DeviceId,
+    pub request_id: RequestId,
     pub msg: T,
 }
 
 impl<T: Message + ClientMessageType + Debug> Message for IdentifiedMessage<T> {
     type Result = T::Result;
 }
 
-#[derive(Debug)]
-pub struct CreateRoom;
-
-impl Message for CreateRoom {
-    type Result = RoomId;
-}
-
-impl ClientMessageType for CreateRoom {}
-
 pub struct ClientServer {
-    sessions: DHashMap<SessionId, Addr<ClientWsSession>>,
-    user_to_sessions: DHashMap<UserId, Vec<SessionId>>,
-    rooms: DHashMap<RoomId, Room>,
+    sessions: DashMap<DeviceId, Addr<ClientWsSession>>,
+    online_devices: DashMap<UserId, Vec<DeviceId>>,
+    rooms: DashMap<RoomId, Vec<UserId>>,
 }
 
 impl ClientServer {
     pub fn new() -> Self {
         ClientServer {
-            sessions: DHashMap::default(),
-            user_to_sessions: DHashMap::default(),
-            rooms: DHashMap::default(),
+            sessions: DashMap::default(),
+            online_devices: DashMap::default(),
+            rooms: DashMap::default(),
+        }
+    }
+
+    fn logout_user_sessions(&mut self, user_id: &UserId) {
+        if let Some(online_devices) = self.online_devices.get(user_id) {
+            online_devices
+                .iter()
+                .map(|id| self.sessions.get_mut(id).unwrap())
+                .for_each(|s| s.do_send(LogoutThisSession));
         }
     }
 
-    fn send_to_room(&mut self, room: &RoomId, message: ServerMessage, sender: &SessionId) {
+    fn logout_sessions(&mut self, logout: LogoutSessions) {
+        // TODO could probably be optimised
+        for (user_id, device_id) in logout.list {
+            if let Some(online_devices) = self.online_devices.get(&user_id) {
+                let device_id = online_devices.iter().find(|id| **id == device_id);
```

Comment:
I think this is just a `.contains`?

Reply:
No, this returns the device id. Contains only checks if it exists.

Reply:
You already have the device id

Reply:
Oh lol